File: src/main/java/com/bn/milk/criteria/DistanceCriteria.java
package com.bn.milk.criteria;

import com.bn.milk.model.Driver;
import com.bn.milk.model.Order;
import com.bn.milk.model.Vehicle;

public class DistanceCriteria extends BaseCriteria {
    public DistanceCriteria(double weight) {
        super("Distance", weight);
    }

    @Override
    public double calculate(Order order, Vehicle vehicle, Driver driver) {
        // Tính khoảng cách từ vị trí xe đến điểm đến của đơn hàng
        double distance = calculateDistance(
                vehicle.getLatitude(), vehicle.getLongitude(),
                order.getLatitudeDest(), order.getLongitudeDest()
        );
        // Normalize distance (0-1, càng gần càng tốt)
        return Math.max(0, 1 - (distance / 1000.0)); // Giả sử max distance = 1000km
    }

    private double calculateDistance(double lat1, double lon1, double lat2, double lon2) {
        final int R = 6371; // Radius of the earth in km
        double latDistance = Math.toRadians(lat2 - lat1);
        double lonDistance = Math.toRadians(lon2 - lon1);
        double a = Math.sin(latDistance / 2) * Math.sin(latDistance / 2)
                + Math.cos(Math.toRadians(lat1)) * Math.cos(Math.toRadians(lat2))
                * Math.sin(lonDistance / 2) * Math.sin(lonDistance / 2);
        double c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
        return R * c;
    }
}

File: backend/src/criteria/DistanceCriteria.ts
// import { BaseCriteria } from './BaseCriteria';
// import { Context } from './Context';

// export class DistanceCriteria extends BaseCriteria {
//   constructor(weight: number) {
//     super('Distance', weight);
//   }

//   // Định nghĩa method calculate đúng kiểu TypeScript
//   calculate(context: Context): number {
   
//     // Chuẩn hóa khoảng cách, giả sử max 1000km
//     return 0;
//   }

//   private calculateDistance(lat1: number, lon1: number, lat2: number, lon2: number): number {
//     const R = 6371; // bán kính Trái đất (km)
//     const dLat = this.degToRad(lat2 - lat1);
//     const dLon = this.degToRad(lon2 - lon1);

//     const a =
//       Math.sin(dLat / 2) * Math.sin(dLat / 2) +
//       Math.cos(this.degToRad(lat1)) * Math.cos(this.degToRad(lat2)) *
//       Math.sin(dLon / 2) * Math.sin(dLon / 2);

//     const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
//     return R * c;
//   }

//   private degToRad(deg: number): number {
//     return deg * (Math.PI / 180);
//   }
// }
